File: Day3/AdventDay3.py
def find_adjacent_symbol(ascii_value):
    """This function checks if the given character (by its ASCII value) 
    is NOT a digit (48-57), a decimal point (.), or a newline character.
    It returns True if the character is something other than these."""
    return(not(48 <= ascii_value <= 57) and ascii_value != 46 and ascii_value != ord('\n'))

if (__name__ == '__main__'):
    # Open the input file "input3.txt" for reading.
    with open("input3.txt", "r") as f:
        # Read all lines from the file into the 'lines' list.
        lines = f.readlines()

    number = final_sum = 0  # Initialize 'number' and 'final_sum' to 0.
    adjacent = False  # Initialize 'adjacent' to False.

    # Loop through each line in the 'lines' list.
    for i in range(len(lines)):
        adjacent = False  # Reset 'adjacent' to False for each new line.
        number = 0  # Reset 'number' to 0 for each new line.

        # Loop through each character in the current line.
        for j in range(len(lines[i])):
            # Check if the current character is a digit (using its ASCII value).
            if 48 <= ord(lines[i][j]) <= 57:
                # If it's a digit, build the 'number' by multiplying the previous 'number' by 10 and adding the current digit. This handles multi-digit numbers.
                number = number * 10 + int(lines[i][j])

                # If 'adjacent' is not yet set to True, check for adjacent symbols.
                if not adjacent:
                    # Check if not the first row.
                    if i != 0:
                        adjacent = find_adjacent_symbol(ord(lines[i-1][j]))
                        # Check the top-left diagonal.
                        if not adjacent and j != 0:
                            adjacent = find_adjacent_symbol(ord(lines[i-1][j-1]))
                        # Check the top-right diagonal.
                        if not adjacent and j != len(lines[i]) - 1:
                            adjacent = find_adjacent_symbol(ord(lines[i-1][j+1]))

                    # Check the symbol directly below the current one (if not the last row).
                    if not adjacent and i != len(lines) - 1:
                        adjacent = find_adjacent_symbol(ord(lines[i+1][j]))
                        # Check the bottom-left diagonal.
                        if not adjacent and j != 0:
                            adjacent = find_adjacent_symbol(ord(lines[i+1][j-1]))
                        # Check the bottom-right diagonal.
                        if not adjacent and j != len(lines[i]) - 1:
                            adjacent = find_adjacent_symbol(ord(lines[i+1][j+1]))

                    # Check the symbol to the right.
                    if not adjacent and j != len(lines[i]) - 1:
                        adjacent = find_adjacent_symbol(ord(lines[i][j+1]))
                    # Check the symbol to the left.
                    if not adjacent and j != 0:
                        adjacent = find_adjacent_symbol(ord(lines[i][j-1]))

            # If the current character is not a digit.
            else:
                # If 'adjacent' is True, add the current 'number' to 'final_sum'.
                if adjacent:
                    final_sum += number  # Add the number to the final sum.


                adjacent = False  # Reset 'adjacent' to False.
                number = 0  # Reset 'number' to 0 to prepare for the next number.

    # After processing all lines and numbers, print the final sum.
    print(final_sum)

File: Day3/AdventDay3.cpp
#include <fstream>
#include <iostream>
#include <string>
#include <cctype>  
#include <vector>
using namespace std;

// Function to check if the character is not a digit, period, or newline
bool find_adjacent_symbol(char ch) {
    return !(isdigit(ch) || ch == '.' || ch == '\n');
}

int main() {
    ifstream file("input3.txt");  // Open the file
    if (!file.is_open()) {
        cerr << "Error opening the file!" << endl;
        return 1;
    }

    string line;
    int number = 0, final_sum = 0;
    bool adjacent = false;

    // Read all lines from the file into a vector for easy access
    vector<string> lines;
    while (getline(file, line)) {
        lines.push_back(line);
    }

    // Iterate through each line and character in the file
    for (int i = 0; i < lines.size(); i++) {
        line = lines[i];
        adjacent = false;
        number = 0;

        for (int j = 0; j < line.length(); j++) {
            if (isdigit(line[j])) {  // If it's a digit, build the number
                number = number * 10 + (line[j] - '0');
                if (!adjacent) {
                    // Check adjacent positions (top, bottom, left, right, diagonals)
                    // Top (i > 0 to ensure we don't go out of bounds)
                    if (i > 0 && find_adjacent_symbol(lines[i - 1][j])) adjacent = true;
                    // Top-left diagonal
                    if (i > 0 && j > 0 && find_adjacent_symbol(lines[i - 1][j - 1])) adjacent = true;
                    // Top-right diagonal
                    if (i > 0 && j < line.length() - 1 && find_adjacent_symbol(lines[i - 1][j + 1])) adjacent = true;

                    // Bottom (i < lines.size() - 1 to ensure we don't go out of bounds)
                    if (i < lines.size() - 1 && find_adjacent_symbol(lines[i + 1][j])) adjacent = true;
                    // Bottom-left diagonal
                    if (i < lines.size() - 1 && j > 0 && find_adjacent_symbol(lines[i + 1][j - 1])) adjacent = true;
                    // Bottom-right diagonal
                    if (i < lines.size() - 1 && j < line.length() - 1 && find_adjacent_symbol(lines[i + 1][j + 1])) adjacent = true;

                    // Left (j > 0 to ensure we don't go out of bounds)
                    if (j > 0 && find_adjacent_symbol(line[j - 1])) adjacent = true;
                    // Right (j < line.length() - 1 to ensure we don't go out of bounds)
                    if (j < line.length() - 1 && find_adjacent_symbol(line[j + 1])) adjacent = true;
                }
            } else {  // If it's not a digit
                if (adjacent) {  // If we found adjacent symbols around the number
                    final_sum += number;
                    cout << "Number found: " << number << endl;
                }
                adjacent = false;
                number = 0;  // Reset the number after each non-digit
            }
        }

        // Handle the case where the line ends with a valid number
        if (adjacent) {
            final_sum += number;
        }
    }

    cout << "Total sum: " << final_sum << endl;
    file.close();  // Close the file

    return 0;
}
